source/agora/consensus/data/Transaction.d

```diff
@@ -48,6 +49,61 @@ public enum TxType : ubyte
 
 public struct Transaction
 {
+    @safe:
+
+    /// ctor with only outputs
+    this (in TxType type, in Output[] outputs) nothrow
+    {
+        this.type = type;
+        this.outputs = outputs.dup; // .sort.array; (FIXME: fix compilation error)
+    }
+
+    /// ctor with only inputs
+    this (in Input[] inputs) nothrow
+    {
+        this.type = TxType.Freeze;  // Can only be freeze
+        this.inputs = inputs.dup.sort.array;
+    }
+
+    /// enable sorting the inputs and outputs on construction
+    this (in TxType type, in Input[] inputs, in Output[] outputs,
+        in Height lock_height = Height(0)) nothrow
+    {
+        this(type, outputs);
+        this.inputs = inputs.dup.sort.array;
+        this.lock_height = lock_height;
+    }
+
+    /// ctor with all params
+    this (in TxType type, in Input[] inputs, in Output[] outputs,
+        inout DataPayload payload, in Height lock_height = Height(0)) nothrow
+    {
+        this(type, inputs, outputs, lock_height);
+        this.payload = payload;
+    }
+
+    /// ctor with all params for immutable
+    this (immutable TxType type, immutable Input[] inputs, immutable Output[] outputs,
+        immutable DataPayload payload, immutable Height lock_height = Height(0)) immutable nothrow
```

Comment:
```suggestion
    /// ctor with all params for immutable
    public this (in TxType type, inout Input[] inputs, inout Output[] outputs,
        inout DataPayload payload, in Height lock_height = Height(0)) inout nothrow
```

Something like this will also support `const`.

Reply:
Updated